cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/CoreSchedulerService.java

```diff
@@ -364,29 +377,41 @@ public void deleteSchedules(Iterable<? extends ScheduleId> scheduleIds) throws N
   @Override
   public void deleteSchedules(ApplicationId appId) {
     checkStarted();
-    execute((StoreAndQueueTxRunnable<Void, RuntimeException>) (store, queue) -> {
-      long deleteTime = System.currentTimeMillis();
-      deleteSchedulesInScheduler(store.listSchedules(appId));
-      List<ScheduleId> deleted = store.deleteSchedules(appId);
-      for (ScheduleId scheduleId : deleted) {
-        queue.markJobsForDeletion(scheduleId, deleteTime);
-      }
-      return null;
-    }, RuntimeException.class);
+    List<ScheduleId> deletedSchedules =
+      execute((StoreAndQueueTxRunnable<List<ScheduleId>, RuntimeException>) (store, queue) -> {
+        long deleteTime = System.currentTimeMillis();
+        List<ProgramSchedule> schedules = store.listSchedules(appId);
+        deleteSchedulesInScheduler(schedules);
+        List<ScheduleId> deleted = store.deleteSchedules(appId);
+        for (ScheduleId scheduleId : deleted) {
+          queue.markJobsForDeletion(scheduleId, deleteTime);
+        }
+        for (ProgramSchedule programSchedule : schedules) {
```

Comment:
is the deleted list ever different from the schedules list?

Reply:
Confirmed with Mao, it should be same

Reply:
if it's the same, I would suggest removing the deleted list because it makes this code more confusing. You can combine this loop with the one directly above it.

Reply:
So I look at the code, it seems like they may be different, because when the store saves trigger information, which is used for reverse indexing, so the deleted list may be larger. So I didn't combine the loop but removed the unnecessary  return of deleted list